cmd/puma-dev/main_test.go

```diff
@@ -1,19 +1,124 @@
 package main
 
 import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
 	"os"
 	"os/exec"
+	"path/filepath"
+	"strings"
 	"testing"
+	"time"
 
+	dev "github.com/puma/puma-dev/dev"
 	. "github.com/puma/puma-dev/dev/devtest"
+
+	"github.com/avast/retry-go"
+	"github.com/puma/puma-dev/homedir"
 	"github.com/stretchr/testify/assert"
 )
 
+var testAppLinkDirPath = homedir.MustExpand("~/.gotest-puma-dev")
+
 func TestMain(m *testing.M) {
 	EnsurePumaDevDirectory()
 	os.Exit(m.Run())
 }
 
+func generateLivePumaDevCertIfNotExist(t *testing.T) {
+	liveSupportPath := homedir.MustExpand(dev.SupportDir)
+	liveCertPath := filepath.Join(liveSupportPath, "cert.pem")
+	liveKeyPath := filepath.Join(liveSupportPath, "key.pem")
+
+	if !FileExists(liveCertPath) || !FileExists(liveKeyPath) {
+		MakeDirectoryOrFail(t, liveSupportPath)
+
+		if err := dev.GeneratePumaDevCertificateAuthority(liveCertPath, liveKeyPath); err != nil {
+			assert.FailNow(t, err.Error())
+		}
+	}
+}
+
+func backgroundPumaDev(t *testing.T) func() {
+	StubCommandLineArgs()
+
+	SetFlagOrFail(t, "dir", testAppLinkDirPath)
+
+	generateLivePumaDevCertIfNotExist(t)
+
+	go func() {
+		main()
+	}()
+
+	count := 1
+	err := retry.Do(func() error {
+		if _, err := net.DialTimeout("tcp", fmt.Sprintf("localhost:%d", *fHTTPPort), time.Duration(10*time.Second)); err != nil {
+			count = count + 1
+			return err
+		}
+		return nil
+	})
+	fmt.Printf("waited %d steps\n", count)
+
+	assert.NoError(t, err)
+
+	return func() {
+		RemoveDirectoryOrFail(t, testAppLinkDirPath)
+	}
+}
+
+func getUrlWithHost(t *testing.T, url string, host string) string {
+	req, _ := http.NewRequest("GET", url, nil)
+	req.Host = host
+
+	resp, err := http.DefaultClient.Do(req)
+
+	if err != nil {
+		assert.FailNow(t, err.Error())
+	}
+
+	defer resp.Body.Close()
+
+	bodyBytes, _ := ioutil.ReadAll(resp.Body)
+
+	return strings.TrimSpace(string(bodyBytes))
+}
+
+func TestMainPumaDev(t *testing.T) {
+	defer backgroundPumaDev(t)()
+
+	projectRoot := Basepath
```

Comment:
don't need the intermediary variable. `Basepath` isn't super clear. `devtest.ProjectRoot` might be better?